mungegithub/mungers/submit-queue.go

```diff
@@ -650,6 +667,89 @@ func (sq *SubmitQueue) updateHealth() {
 	}
 }
 
+func (sq *SubmitQueue) getRunningPRNumber() int {
+	defer sq.Unlock()
+	sq.Lock()
+	if sq.githubE2ERunning != nil {
+		return *sq.githubE2ERunning.Issue.Number
+	}
+	// make sure to return an invalid pr number otherwise
+	return -1
+}
+
+func (sq *SubmitQueue) monitorProw() {
+	nonBlockingJobNames := make(map[string]bool)
+	for _, jobName := range sq.NonBlockingJobNames {
+		nonBlockingJobNames[jobName] = true
+	}
+	requireRetestJobNames := make(map[string]bool)
+	for _, jobName := range mungeopts.RequiredContexts.Retest {
+		requireRetestJobNames[jobName] = true
+	}
+	url := sq.ProwURL + "/data.js"
+	for {
+		currentPR := sq.getRunningPRNumber()
+		lastPR := sq.githubE2ELastPRNum
+		// get current job info from prow
+		allJobs, err := getJobs(url)
+		if err != nil {
+			glog.Errorf("Error reading batch jobs from Prow URL %v: %v", url, err)
+			continue
+		}
+		// TODO: copy these from sq first instead
+		ciStatus := make(map[string]map[string]jobStatus)
+		ciLatest := make(map[string]map[string]time.Time)
+
+		for _, job := range allJobs {
+			if job.Finished == "" || job.BuildID == "" {
+				continue
+			}
+			// type/category
+			key := job.Type + "/"
+			// the most recent submit-queue PR(s)
+			if job.Number == currentPR || job.Number == lastPR {
+				key += "serial"
+			} else if nonBlockingJobNames[job.Job] {
+				key += "nonblocking"
+			} else if requireRetestJobNames[job.Job] {
+				key += "requiredretest"
+			}
+
+			ft, err := time.Parse(time.RFC3339Nano, job.Finished)
+			if err != nil {
+				glog.Errorf("Error parsing job finish time %s: %v", job.Finished, err)
+				continue
+			}
+
+			if _, ok := ciLatest[key]; !ok {
+				ciLatest[key] = make(map[string]time.Time)
+				ciStatus[key] = make(map[string]jobStatus)
+			}
+			latest, ok := ciLatest[key][job.Job]
+
+			// TODO: flake cache?
+			if !ok || latest.After(ft) {
+				ciLatest[key][job.Job] = ft
+				ciStatus[key][job.Job] = jobStatus{
+					State:   job.State,
+					BuildID: job.BuildID,
+					URL:     job.URL,
+				}
+			}
+		}
+
+		sq.setCIStatus(ciStatus)
+
+		time.Sleep(1 * time.Minute)
```

Comment:
ok, this pattern is common throughout mungegithub. should we fix the others at some point?